kakao/src/main/kotlin/com/agoda/kakao/delegate/ViewInteractionDelegate.kt

```diff
@@ -0,0 +1,47 @@
+package com.agoda.kakao.delegate
+
+import androidx.test.espresso.*
+import com.agoda.kakao.Kakao
+import com.agoda.kakao.intercept.Interceptor
+import com.agoda.kakao.screen.Screen
+import org.hamcrest.Matcher
+import org.jetbrains.annotations.Contract
+
+/**
+ * Delegation class for [ViewInteraction].
+ * Wraps all available public calls and intercepts [check] and [perform].
+ *
+ * @see Delegate
+ * @see Interceptor
+ */
+class ViewInteractionDelegate(override var interaction: ViewInteraction) : Delegate<ViewInteraction, ViewAssertion, ViewAction> {
+    override var interceptor: Interceptor<ViewInteraction, ViewAssertion, ViewAction>? = null
+
+    @Contract("_->this")
+    fun check(viewAssert: ViewAssertion) = this.also {
+        if (!interceptCheck(viewAssert)) interaction.check(viewAssert)
+    }
+
+    @Contract("_->this")
+    fun inRoot(rootMatcher: Matcher<Root>) = this.also {
+        interaction.inRoot(rootMatcher)
+    }
+
+    @Contract("_->this")
+    fun noActivity() = this.also {
+        interaction.noActivity()
+    }
+
+    @Contract("_->this")
+    fun perform(vararg viewActions: ViewAction) = this.also {
+        if (viewActions.none { interceptPerform(it) }) interaction.perform(*viewActions)
+    }
+
+    @Contract("_->this")
+    fun withFailureHandler(failureHandler: FailureHandler) = this.also {
+        interaction.withFailureHandler(failureHandler)
+    }
+
+    override fun screenInterceptor() = Screen.vis.takeIf { it.isNotEmpty() }?.peek()
```

Comment:
1) screenInterceptor and kakaoInterceptor implementations are the same in view*, data* and web interaction delegates. Maybe it can be moved to Delegate interface?
2) Why only interceptors of the latest screen that became active are used? Maybe interceptors of all nested screens should be called sequentially?

Reply:
1. why not
2. agree

Reply:
1. They reference different stacks/holders since the interceptor types are different. I'm not sure how can we abstract that, since we need to reference the interceptor matching to the type of delegate, and delegate interface does not hold any type information.

Reply:
2. Maybe